File: filesystem/src/test/scala/net/kemitix/thorp/filesystem/MD5HashGeneratorTest.scala
package net.kemitix.thorp.filesystem

import java.nio.file.Path

import net.kemitix.thorp.domain.MD5HashData.{BigFile, Root}
import org.scalatest.FunSpec
import zio.DefaultRuntime

class MD5HashGeneratorTest extends FunSpec {

  describe("md5File()") {
    describe("read a small file (smaller than buffer)") {
      val path = Resource(this, "upload/root-file").toPath
      it("should generate the correct hash") {
        val expected = Right(Root.hash)
        val result   = invoke(path)
        assertResult(expected)(result)
      }
    }

    describe("read a large file (bigger than buffer)") {
      val path = Resource(this, "big-file").toPath
      it("should generate the correct hash") {
        val expected = Right(BigFile.hash)
        val result   = invoke(path)
        assertResult(expected)(result)
      }
    }

    def invoke(path: Path) =
      new DefaultRuntime {}.unsafeRunSync {
        MD5HashGenerator
          .md5File(path)
          .provide(testEnv)
      }.toEither
  }

  describe("md5FileChunk") {
    describe("read chunks of file") {
      val path = Resource(this, "big-file").toPath
      it("should generate the correct hash for first chunk of the file") {
        val expected = Right(BigFile.Part1.hash.hash())
        val result =
          invoke(path, BigFile.Part1.offset, BigFile.Part1.size).map(x =>
            x.hash)
        assertResult(expected)(result)
      }
      it("should generate the correct hash for second chunk of the file") {
        val expected = Right(BigFile.Part2.hash.hash())
        val result =
          invoke(path, BigFile.Part2.offset, BigFile.Part2.size).map(x =>
            x.hash)
        assertResult(expected)(result)
      }
    }

    def invoke(path: Path, offset: Long, size: Long) =
      new DefaultRuntime {}.unsafeRunSync {
        MD5HashGenerator
          .md5FileChunk(path, offset, size)
          .provide(testEnv)
      }.toEither
  }

  type TestEnv = FileSystem
  val testEnv: TestEnv = new FileSystem.Live {}

}

File: filesystem/src/test/java/net/kemitix/thorp/filesystem/MD5HashGeneratorTest.java
package net.kemitix.thorp.filesystem;

import net.kemitix.thorp.domain.MD5Hash;
import net.kemitix.thorp.domain.MD5HashData;
import org.assertj.core.api.WithAssertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.file.Path;
import java.security.NoSuchAlgorithmException;

public class MD5HashGeneratorTest
        implements WithAssertions {
    @Nested
    @DisplayName("md5File")
    public class Md5File {
        @Test
        @DisplayName("read a file smaller than buffer")
        public void readSmallFile() throws IOException, NoSuchAlgorithmException {
            Path path = Resource.select(this, "upload/root-file").toPath();
            MD5Hash result = MD5HashGenerator.md5File(path);
            assertThat(result).isEqualTo(MD5HashData.Root.hash);
        }
        @Test
        @DisplayName("read a file larger than buffer")
        public void readLargeFile() throws IOException, NoSuchAlgorithmException {
            Path path = Resource.select(this, "big-file").toPath();
            MD5Hash result = MD5HashGenerator.md5File(path);
            assertThat(result).isEqualTo(MD5HashData.BigFile.hash);
        }
    }
    @Nested
    @DisplayName("md5FileChunk")
    public class Md5FileChunk {
        @Test
        @DisplayName("read first chunk of file")
        public void chunk1() throws IOException, NoSuchAlgorithmException {
            Path path = Resource.select(this, "big-file").toPath();
            MD5Hash result = MD5HashGenerator.md5FileChunk(path,
                    MD5HashData.BigFile.Part1.offset,
                    MD5HashData.BigFile.Part1.size);
            assertThat(result).isEqualTo(MD5HashData.BigFile.Part1.hash);
        }
        @Test
        @DisplayName("read second chunk of file")
        public void chunk2() throws IOException, NoSuchAlgorithmException {
            Path path = Resource.select(this, "big-file").toPath();
            MD5Hash result = MD5HashGenerator.md5FileChunk(path,
                    MD5HashData.BigFile.Part2.offset,
                    MD5HashData.BigFile.Part2.size);
            assertThat(result).isEqualTo(MD5HashData.BigFile.Part2.hash);
        }
    }
}
